src/main/java/org/whispersystems/textsecuregcm/storage/DirectoryReconciler.java

```diff
@@ -0,0 +1,248 @@
+        workerCount = Math.max(reconciliationCache.getWorkerCount(WORKER_TTL_MS), 1);
+      }
+    } catch (Exception ex) {
+      logger.warn("error in directory reconciliation", ex);
+    }
+
+    if (!accelerate) {
+      scheduleWithJitter(CHUNK_INTERVAL * workerCount);
+    } else {
+      scheduleWithJitter(ACCELERATE_CHUNK_INTERVAL);
+    }
+  }
+
+  private void scheduleWithJitter(long delayMs) {
+    long randomJitterMs = (long) (random.nextDouble() * JITTER_MAX * delayMs);
+    try {
+      executor.schedule(this, delayMs + randomJitterMs, TimeUnit.MILLISECONDS);
+    } catch (RejectedExecutionException ex) {
+      logger.info("reconciler shutting down");
+    }
+  }
+
+  private boolean processChunk() {
+    Optional<String>               fromNumber          = reconciliationCache.getLastNumber();
+    int                            chunkSize           = (int) Math.max(getAccountCount() * PERIOD / CHUNK_INTERVAL, MINIMUM_CHUNK_SIZE);
+    DirectoryReconciliationRequest request             = readChunk(fromNumber, chunkSize);
+    Response                       sendChunkResponse   = sendChunk(request);
+    boolean                        sendChunkSuccessful = sendChunkResponse.getStatusInfo().getFamily() == Response.Status.Family.SUCCESSFUL;
+
+    if (sendChunkResponse.getStatus() == 404 || request.getToNumber() == null) {
+      reconciliationCache.clearAccelerate();
+    }
+
+    if (sendChunkSuccessful) {
+      reconciliationCache.setLastNumber(Optional.fromNullable(request.getToNumber()));
+    } else if (sendChunkResponse.getStatus() == 404) {
+      reconciliationCache.setLastNumber(Optional.absent());
+    }
+
+    return sendChunkSuccessful;
+  }
+
+  private DirectoryReconciliationRequest readChunk(Optional<String> fromNumber, int chunkSize) {
+    try (Timer.Context timer = readChunkTimer.time()) {
+      List<Account> accounts = accountsManager.getAllFrom(fromNumber, chunkSize);
+
+      List<String> numbers = new ArrayList<>(accounts.size());
+      for (Account account : accounts) {
+        if (account.isActive()) {
+          numbers.add(account.getNumber());
+        }
+      }
+
+      Optional<String> toNumber = Optional.absent();
+      if (!accounts.isEmpty()) {
+        toNumber = Optional.of(accounts.get(accounts.size() - 1).getNumber());
+      }
+
+      return new DirectoryReconciliationRequest(fromNumber.orNull(), toNumber.orNull(), numbers);
+    }
+  }
+
+  private Response sendChunk(DirectoryReconciliationRequest request) {
+    try (Timer.Context timer = sendChunkTimer.time()) {
+      Response response = reconciliationClient.sendChunk(request);
+      if (response.getStatusInfo().getFamily() != Response.Status.Family.SUCCESSFUL) {
+        sendChunkErrorMeter.mark();
+        logger.warn("http error " + response.getStatus());
+      }
+      return response;
+    } catch (ProcessingException ex) {
+      sendChunkErrorMeter.mark();
+      logger.warn("request error: ", ex);
+      throw ex;
```

Comment:
I generally try to avoid re-throwing unwrapped exceptions

Reply:
fixed in 8ed0058f3703.